01 Importing Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

Importing Data

In [2]:
path = r'C:\Users\mngit\OneDrive\Data Analytics\Project Excercises\Achievement 4\Instacart Basket Analysis' #creating path

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data','Original Data', 'products.csv'), index_col = False ) #creating dataframe products.csv

In [4]:
dfords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'), index_col = False ) #creating dataframe orders_wrangled

In [5]:
#create dataframe
df_test = pd.DataFrame()

In [6]:
#create a mixed type column
df_test['mix'] = ['a','b',1,True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#checking for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_prods.isnull().sum() #check if there are any missing values

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull()==True] #finding missing values

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False] #creating subset excluding missing values

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]  #looking for full duplicates

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates() #creating new dataframe that doesn’t include the duplicates

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Observations requiring further investigation on the cleaned Products dataframe
1. Min value is 1 for all variabes
2. Max price is much higher than mean and std deviation

Task 4.4 - Answers

In [21]:
#checking for mixed types

for col in dfords.columns.tolist():
  weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (dfords[weird]) > 0:
    print (col)


There are no mixed type data on the df_ords dataframe

In [22]:
dfords.isnull().sum() #check if there are any missing values

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

206,209 values are missing from the days_since_prior_order variable

In [23]:
df_nan_dfords = dfords[dfords['days_since_prior_order'].isnull()==True]#creating subset of dataframe containing only values that have missing data 

In [24]:
df_nan_dfords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

There are 206,209 missing values from the days_since_prior_order. All missing values also have the same order number (1) which could imply that this data was never captured/collected

In [25]:
dfords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [26]:
dfords.median()# checking for median value

C:\Users\mngit\AppData\Local\Temp\ipykernel_10728\2421960545.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  dfords.median()# checking for median value


Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [27]:
 dfords['days_since_prior_order'].fillna(7, inplace=True) #imputing median value(7) fo missing days_since_prior_order 

Chose to impute and not delete records with missing values as there is a substantial number of these records. 

Used Median value (7) as opposed to the mean to avoid over-weighting one side


In [28]:
dfords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [29]:
dfords_clean = dfords[dfords['days_since_prior_order'].isnull() == False]

In [30]:
df_dups = dfords_clean[dfords_clean.duplicated()] #checking for duplicates

In [31]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicate values in the orders dataframe

In [32]:
dfords_clean.shape

(3421083, 8)

In [33]:
 #exporting cleaned orders data frame
dfords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv')) #exporting cleaned products data frame

In [35]:
df_prods_clean_no_dups.shape

(49672, 5)